File: OpenWrap/Basic/Mopub/ObjectiveC/SampleApp/SampleApp/EventHandlers/MoPub/Interstitial/POBInterstitialCustomEvent.m
// This class is compatible with OpenWrap SDK v1.5.0 and MoPub SDK 5.10.0

#import "POBInterstitialCustomEvent.h"
#import <POBBid.h>
#import <POBRenderer.h>

@interface POBInterstitialCustomEvent () <POBInterstitialRendererDelegate>
@property (nonatomic, strong) id<POBInterstitialRendering> currentRenderer;
@property (nonatomic, weak) UIViewController *viewController;
@end

@implementation POBInterstitialCustomEvent

// Method required for MoPub SDK versions 5.9.0 and below
- (void)requestInterstitialWithCustomEventInfo:(NSDictionary *)info {
    [self loadAdWithCustomEventInfo:info];
}

// Method required for MoPub SDK versions 5.10.0 and above
- (void)requestInterstitialWithCustomEventInfo:(NSDictionary *)info adMarkup:(NSString *)adMarkup {
    [self loadAdWithCustomEventInfo:info];
}

- (void)showInterstitialFromRootViewController:(UIViewController *)rootViewController {
    self.viewController = rootViewController;
    NSDictionary *customData = self.localExtras[@"customData"];
    NSNumber *loadTimeOrientation = customData[@"loadTimeOrientation"];
    [_currentRenderer showFromViewController:_viewController
                               inOrientation:loadTimeOrientation.integerValue];
}

#pragma mark - Supporting Methods

- (void)loadAdWithCustomEventInfo:(NSDictionary *)info {
    if (self.localExtras == nil) {
        [self handleCustomEventFailure];
        return;
    }
    POBBid *bid = self.localExtras[@"pob_bid"];
    bid.hasWon = YES;
    _currentRenderer = [POBRenderer interstitialRenderer];
    [_currentRenderer setDelegate:self];
    [_currentRenderer renderAdDescriptor:bid];
}

-(void)handleCustomEventFailure{
    NSError *error = [NSError errorWithDomain:kPOBErrorDomain code:POBErrorNoAds userInfo:@{NSLocalizedDescriptionKey:NSLocalizedString(@"Bid not available for custom event", comment: @"Bid not available for custom event")}];
    [self.delegate interstitialCustomEvent:self didFailToLoadAdWithError:error];
}

#pragma mark - PMInterstitialRendererDelegate

- (void)interstitialRendererDidRenderAd {
    [self.delegate interstitialCustomEvent:self didLoadAd:[UIView new]];
}

- (void)interstitialRendererDidFailToRenderAdWithError:(NSError *)error {
    [self.delegate interstitialCustomEvent:self didFailToLoadAdWithError:error];
}

- (void)interstitialRendererDidClick {
    [self.delegate interstitialCustomEventDidReceiveTapEvent:self];
}

- (void)interstitialRendererWillLeaveApp {
    [self.delegate interstitialCustomEventWillLeaveApplication:self];
}

- (void)interstitialRendererWillPresentModal {
    [self.delegate interstitialCustomEventWillAppear:self];
    [self.delegate interstitialCustomEventDidAppear:self];
}

- (void)interstitialRendererDidDismissModal {
    [self.delegate interstitialCustomEventWillDisappear:self];
    [self.delegate interstitialCustomEventDidDisappear:self];
}

- (void)interstitialRendererDidExpireAd {
    [self.delegate interstitialCustomEventDidExpire:self];
}

- (UIViewController *)viewControllerForPresentingModal {
    return _viewController;
}

- (void)dealloc {
    _currentRenderer = nil;
}

@end

File: OpenWrap/Basic/Mopub/Swift/SampleApp/SampleApp/EventHandlers/MoPub/Interstitial/POBInterstitialCustomEvent.swift
// This class is compatible with OpenWrap SDK v1.5.0 and MoPub SDK 5.10.0

import UIKit

class POBInterstitialCustomEvent: MPInterstitialCustomEvent, POBInterstitialRendererDelegate {
    
    var currentRenderer: POBInterstitialRendering?
    weak var viewController: UIViewController?
    
    // Method required for MoPub SDK versions 5.9.0 and below
    override func requestInterstitial(withCustomEventInfo info: [AnyHashable : Any]!) {
        self.loadAd(withCustomEventInfo: info)
    }
    
    // Method required for MoPub SDK versions 5.10.0 and above
    override func requestInterstitial(withCustomEventInfo info: [AnyHashable : Any]!, adMarkup: String!) {
        self.loadAd(withCustomEventInfo: info)
    }
    
    override func showInterstitial(fromRootViewController rootViewController: UIViewController!) {
        if rootViewController != nil {
            viewController = rootViewController
            let customData : [String:Any] = localExtras?["customData"] as! [String : Any]
            let loadTimeOrientation : NSNumber = customData["loadTimeOrientation"] as! NSNumber
            currentRenderer?.show(from: viewController!, in: UIInterfaceOrientation(rawValue:loadTimeOrientation.intValue)!)
        }
    }

    // MARK: - Supporting Methods
    
    func loadAd(withCustomEventInfo info: [AnyHashable : Any]!) {
        if localExtras == nil {
            self.handleCustomEventFailure()
            return
        }
        if let bid = localExtras["pob_bid"] as? POBBid {
            bid.hasWon = true
            currentRenderer = POBRenderer.interstitialRenderer()
            currentRenderer?.setDelegate(self)
            currentRenderer?.renderAdDescriptor(bid)
        }else{
            self.handleCustomEventFailure()
        }
    }

    func handleCustomEventFailure() -> Void {
        self.delegate.interstitialCustomEvent(self, didFailToLoadAdWithError: NSError(domain: kPOBErrorDomain, code: POBErrorCode.errorInternalError.rawValue, userInfo: [NSLocalizedDescriptionKey : NSLocalizedString("Bid not available for custom event", comment: "Bid not available for custom event")]))
    }

// MARK: - POBInterstitialRendererDelegate

    func interstitialRendererDidRenderAd() {
        delegate.interstitialCustomEvent(self, didLoadAd: UIView())
    }
    
    func interstitialRendererDidFailToRenderAdWithError(_ error: Error!) {
        delegate.interstitialCustomEvent(self, didFailToLoadAdWithError: error)
    }
    
    func interstitialRendererDidClick() {
        delegate.interstitialCustomEventDidReceiveTap(self)
    }
    
    func interstitialRendererWillLeaveApp() {
        delegate.interstitialCustomEventWillLeaveApplication(self)
    }
    
    func interstitialRendererWillPresentModal() {
        delegate.interstitialCustomEventWillAppear(self)
        delegate.interstitialCustomEventDidAppear(self)
    }
    
    func interstitialRendererDidDismissModal() {
        delegate.interstitialCustomEventWillDisappear(self)
        delegate.interstitialCustomEventDidDisappear(self)
    }
    
    func interstitialRendererDidExpireAd() {
        delegate.interstitialCustomEventDidExpire(self)
    }
    
    func viewControllerForPresentingModal() -> UIViewController! {
        return viewController!
    }
    
    deinit {
        currentRenderer = nil
    }
}
